// ra-core/lib/sideEffect/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.undoSaga = exports.i18nSaga = exports.refreshSaga = exports.accumulateSaga = exports.redirectionSaga = exports.notificationSaga = exports.errorSaga = exports.fetchSaga = exports.authSaga = exports.adminSaga = undefined;

var _admin = require('./admin');

var _admin2 = _interopRequireDefault(_admin);

var _auth = require('./auth');

var _auth2 = _interopRequireDefault(_auth);

var _fetch = require('./fetch');

var _fetch2 = _interopRequireDefault(_fetch);

var _error = require('./error');

var _error2 = _interopRequireDefault(_error);

var _notification = require('./notification');

var _notification2 = _interopRequireDefault(_notification);

var _redirection = require('./redirection');

var _redirection2 = _interopRequireDefault(_redirection);

var _accumulate = require('./accumulate');

var _accumulate2 = _interopRequireDefault(_accumulate);

var _refresh = require('./refresh');

var _refresh2 = _interopRequireDefault(_refresh);

var _i18n = require('./i18n');

var _i18n2 = _interopRequireDefault(_i18n);

var _undo = require('./undo');

var _undo2 = _interopRequireDefault(_undo);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.adminSaga = _admin2.default;
exports.authSaga = _auth2.default;
exports.fetchSaga = _fetch2.default;
exports.errorSaga = _error2.default;
exports.notificationSaga = _notification2.default;
exports.redirectionSaga = _redirection2.default;
exports.accumulateSaga = _accumulate2.default;
exports.refreshSaga = _refresh2.default;
exports.i18nSaga = _i18n2.default;
exports.undoSaga = _undo2.default;
